Structure-structure coupling

Alex-tru opened the pull request:
This extension of the adapter allows to read displacements and write forces to calculix participants. Should maintain unaltered FSI functionality, though it has not been tested yet.

arusch commented:
Hi Àlex,

thank you for this pull request. I will review the changes as quickly as possible but probably not before the end of this week.

Cheers
Alexander

Alex-tru commented:
Ok, thanks in advance!

Alex-tru commented:
Hi Alexander,
thanks a lot for the tips and recommendations. Let me comment on the points you mentioned:
1. Restored.
2, 3, 4. I think that most for loops with different style were the ones for debug-output, which I've removed now. For the rest, style should now be fitting in.
5. No, there's no reason. A comment-typo.
6. This has a reason. In `CCXHelpers.c`, in the function `getXbounIndices` I introduced a switch statement using the `CouplingDataType` as case variable. It is not strictly necessary, but otherwise an additional flag has to be used.

arusch commented on Makefile:
```diff
@@ -106,4 +106,4 @@ $(OBJDIR):
 
 clean:
-	rm -f $(OBJDIR)/*.o $(OBJDIR)/ccx_2.13.a $(OBJDIR)/ccx_preCICE
+rm -f $(OBJDIR)/*.o $(OBJDIR)/ccx_2.13.a $(OBJDIR)/ccx_preCICE
```
Watch out: Indentation in `Makefiles` is crucial. Deleting the tab here creates a `Makefile` that is not functioning. Please redo this such that there are no changes at the `Makefile` at all.

arusch commented on adapter/PreciceInterface.c:
```diff
@@ -188,18 +187,18 @@ void Precice_ReadCouplingData( SimulationData * sim )
+	    	case DISPLACEMENTDELTAS:
                 printf( "DisplacementDeltas cannot be used as read data\n" );
-                fflush( stdout );
+				fflush( stdout );
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

arusch commented on adapter/PreciceInterface.c:
```diff
@@ -188,18 +187,18 @@ void Precice_ReadCouplingData( SimulationData * sim )
-                exit( EXIT_FAILURE );
-                break;
-	    case DISPLACEMENTDELTAS:
+				// Read and set displacements as single point constraints (Dirichlet BC)
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

arusch commented on adapter/PreciceInterface.c:
```diff
@@ -280,14 +279,13 @@ void Precice_WriteCouplingData( SimulationData * sim )
                 precicec_writeBlockVectorData( interfaces[i]->displacementsDataID, interfaces[i]->numNodes, interfaces[i]->preciceNodeIDs, interfaces[i]->nodeVectorData );
                 break;
-	    case DISPLACEMENTDELTAS:
+	    	case DISPLACEMENTDELTAS:
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

arusch commented on adapter/PreciceInterface.c:
```diff
@@ -330,13 +328,13 @@ void PreciceInterface_Create( PreciceInterface * interface, SimulationData * sim
 	interface->triangles = NULL;
 	interface->nodeScalarData = NULL;
-        interface->nodeVectorData = NULL;
+    interface->nodeVectorData = NULL;
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

arusch commented on adapter/PreciceInterface.c:
```diff
@@ -469,17 +467,26 @@ void PreciceInterface_ConfigureCouplingData( PreciceInterface * interface, Simul
+        	interface->readData = DISPLACEMENTS;
+			interface->xbounIndices = malloc( interface->numNodes * 3 * sizeof( int ) );
+            interface->displacementsDataID = precicec_getDataID( config->readDataNames[i], interface->nodesMeshID );
+			getXbounIndices( interface->nodeIDs, interface->numNodes, sim->nboun, sim->ikboun, sim->ilboun, interface->xbounIndices, DISPLACEMENTS );
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

arusch commented on adapter/PreciceInterface.c:
```diff
@@ -515,20 +522,27 @@ void PreciceInterface_ConfigureCouplingData( PreciceInterface * interface, Simul
-                        interface->displacementsDataID = precicec_getDataID( config->writeDataNames[i], interface->nodesMeshID );
-                        printf( "Write data '%s' found.\n", config->writeDataNames[i] );
-                }
+        else if ( strcmp1( config->writeDataNames[i], "Displacements" + i ) == 0 )
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

arusch commented on adapter/PreciceInterface.c:
```diff
@@ -587,9 +601,8 @@ void PreciceInterface_FreeData( PreciceInterface * preciceInterface )
-    	if ( preciceInterface->xforcIndices != NULL ) {
-        	free( preciceInterface->xforcIndices );
-    	}
+    if ( preciceInterface->xforcIndices != NULL )
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

arusch commented on adapter/PreciceInterface.h:
```diff
@@ -23,7 +23,7 @@
  *  DisplacementDeltas - FSI data to be written (by the Calculix adapter)
  */
-enum CouplingDataType {TEMPERATURE, HEAT_FLUX, CONVECTION, FORCES, DISPLACEMENTS, DISPLACEMENTDELTAS};
+//enum CouplingDataType {TEMPERATURE, HEAT_FLUX, CONVECTION, FORCES, DISPLACEMENTS, DISPLACEMENTDELTAS};
```
Your change concerning the position of the `CouplingDataType` enumeration is confirmed, but please also delete the deprecated code here and don't leave it commented out.

arusch commented on adapter/PreciceInterface.h:
```diff
@@ -55,7 +55,7 @@ typedef struct PreciceInterface {
 	// Arrays to store the coupling data
 	double * nodeScalarData;
-        double * nodeVectorData; //Forces, displacements and displacementDeltas are vector quantities
+    double * nodeVectorData; //Forces, displacements and displacementDeltas are vector quantities
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

arusch commented on adapter/PreciceInterface.h:
```diff
@@ -101,11 +101,11 @@ typedef struct SimulationData {
 	double * co;
 	ITG nboun;
-        ITG nforc; //total number of forces
+    ITG nforc; //total number of forces
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

arusch commented on nonlingeo_precice.c:
```diff
@@ -221,6 +221,7 @@ void nonlingeo_precice(double **cop, ITG *nk, ITG **konp, ITG **ipkonp, char **l
       .xboun = xboun,
       .ntmat_ = ntmat_,
       .vold = vold,
+	  .fn = fn,
```
White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).
